src/bio/bio_bulk.c

```diff
@@ -0,0 +1,723 @@
+	return rc;
+}
+
+static int
+bulk_grp_grow(struct bio_dma_buffer *bdb, struct bio_bulk_group *bbg,
+	      struct bio_bulk_args *arg)
+{
+	struct bio_dma_chunk	*chk;
+	int			 rc;
+
+	/* Try grab an idle chunk first */
+	if (!d_list_empty(&bdb->bdb_idle_list))
+		goto populate;
+
+	/* Grow DMA buffer when not reaching DMA upper bound */
+	if (bdb->bdb_tot_cnt < bio_chk_cnt_max) {
+		rc = dma_buffer_grow(bdb, 1);
+		if (rc != 0)
+			return rc;
+	}
+
+	/* Try to evict an unused chunk from other bulk group */
+	rc = bulk_reclaim_chunk(bdb, bbg);
+	if (rc)
+		return rc;
+
+populate:
+	D_ASSERT(!d_list_empty(&bdb->bdb_idle_list));
+
+	chk = d_list_entry(bdb->bdb_idle_list.next,
+			   struct bio_dma_chunk, bdc_link);
+	rc = bulk_chunk_populate(chk, bbg, arg);
+	if (rc)
+		return rc;
+
+	d_list_move_tail(&chk->bdc_link, &bbg->bbg_dma_chks);
+	bbg->bbg_chk_cnt++;
+
+	return 0;
+}
+
+static void
+bulk_hdl_unhold(struct bio_bulk_hdl *hdl)
+{
+	struct bio_dma_chunk	*chk = hdl->bbh_chunk;
+	struct bio_bulk_group	*bbg;
+
+	D_ASSERT(bulk_hdl_is_inuse(hdl));
+
+	hdl->bbh_inuse = false;
+	hdl->bbh_bulk_off = 0;
+
+	D_ASSERT(chk != NULL);
+	D_ASSERT(chk->bdc_bulk_idle < chk->bdc_bulk_cnt);
+	chk->bdc_bulk_idle++;
+
+	bbg = chk->bdc_bulk_grp;
+	D_ASSERT(bbg != NULL);
+	d_list_add_tail(&hdl->bbh_link, &bbg->bbg_idle_bulks);
+}
+
+static void
+bulk_hdl_hold(struct bio_bulk_hdl *hdl, unsigned int pg_off)
+{
+	struct bio_dma_chunk	*chk = hdl->bbh_chunk;
+
+	D_ASSERT(!bulk_hdl_is_inuse(hdl));
+
+	d_list_del_init(&hdl->bbh_link);
+	hdl->bbh_inuse = true;
+	hdl->bbh_bulk_off = pg_off;
+
+	D_ASSERT(chk != NULL);
+	D_ASSERT(chk->bdc_bulk_idle > 0);
+	chk->bdc_bulk_idle--;
+}
+
+static struct bio_bulk_hdl *
+bulk_get_hdl(struct bio_desc *biod, unsigned int pg_cnt,
+	     unsigned int pg_off, struct bio_bulk_args *arg)
+{
+	struct bio_dma_buffer	*bdb = iod_dma_buf(biod);
+	struct bio_bulk_group	*bbg;
+	struct bio_bulk_hdl	*hdl;
+	int			 rc;
+
+	bbg = bulk_grp_get(bdb, pg_cnt);
+	if (bbg == NULL) {
+		D_ERROR("Failed to get bulk group (%u pages)\n", pg_cnt);
+		dump_dma_info(bdb);
+
+		biod->bd_retry = 1;
+		return NULL;
+	}
+
+	if (!d_list_empty(&bbg->bbg_idle_bulks))
+		goto done;
+
+	rc = bulk_grp_grow(bdb, bbg, arg);
+	if (rc) {
+		D_ERROR("Failed to grow bulk grp (%u pages) "DF_RC"\n",
+			pg_cnt, DP_RC(rc));
+		dump_dma_info(bdb);
+
+		if (rc == -DER_AGAIN)
+			biod->bd_retry = 1;
+
+		return NULL;
+	}
+done:
+	D_ASSERT(!d_list_empty(&bbg->bbg_idle_bulks));
+	hdl = d_list_entry(bbg->bbg_idle_bulks.next, struct bio_bulk_hdl,
+			   bbh_link);
+
+	bulk_hdl_hold(hdl, pg_off);
+	return hdl;
+}
+
+static inline bool
+bypass_bulk_cache(struct bio_iov *biov, unsigned int pg_cnt)
+{
+	/* Hole, no RDMA */
+	if (bio_addr_is_hole(&biov->bi_addr))
+		return true;
+	/* SCM, RDMA to SCM directly */
+	if (biov->bi_addr.ba_type == DAOS_MEDIA_SCM)
+		return true;
+	/* Huge IOV, allocate DMA buffer & create bulk handle on-the-fly */
+	if (pg_cnt > bio_chk_sz)
```

Comment:
large IO's one segment/iov maybe can > 20MB normally?
if that is true, maybe can consider to use multiple chunks for one bulk handle, if it is not too complex?(later)

and for small fragmented IO, like 8 segments each with 8 bytes, seems in that case each 8B segment will use at least one page buffer's bulk handle? just confirm.
or maybe can bypass those very small IOV?

Reply:
Any IOV size larger than chunk size (8MB) is considered as huge IOV, and we don't use DMA chunk for huge IOV at this moment. I agree that we can consider to use multiple DMA chunks & cached bulk handles for the huge IOV later.

For small segment (less than 4k), we have to use a full 4k buffer, because the SPDK IO granularity is usually 4k.